code/go/0chain.net/smartcontract/dbs/event/allocation_blobber_term.go

```diff
@@ -0,0 +1,134 @@
+package event
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"gorm.io/gorm"
+)
+
+type AllocationBlobberTerm struct {
+	gorm.Model
+	AllocationID            string        `json:"allocation_id" gorm:"primaryKey; not null"` // Foreign Key
+	BlobberID               string        `json:"blobber_id" gorm:"primaryKey; not null"`    // Foreign Key
+	ReadPrice               int64         `json:"read_price"`
+	WritePrice              int64         `json:"write_price"`
+	MinLockDemand           float64       `json:"min_lock_demand"`
+	MaxOfferDuration        time.Duration `json:"max_offer_duration"`
+	ChallengeCompletionTime time.Duration `json:"challenge_completion_time"`
+	Allocation              Allocation    `json:"-" gorm:"references:AllocationID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
+	Blobber                 Blobber       `json:"-" gorm:"references:BlobberID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE"`
+}
+
+func (edb *EventDb) GetAllocationBlobberTerm(allocationID string, blobberID string) (*AllocationBlobberTerm, error) {
+	if allocationID == "" && blobberID == "" {
+		return nil, fmt.Errorf("can not retrieve term with empty Allocation ID and blobber ID")
+	}
+	var term AllocationBlobberTerm
+	if err := edb.Store.Get().Model(&AllocationBlobberTerm{}).
+		Where(&AllocationBlobberTerm{AllocationID: allocationID, BlobberID: blobberID}).
+		Scan(&term).Error; err != nil {
+		return nil, fmt.Errorf("failed to retrieve term: AllocationID %v and blobberID %v does not exist, error %v",
+			allocationID, blobberID, err)
+	}
+
+	return &term, nil
+}
+
+func (edb *EventDb) GetAllocationBlobberTerms(allocationID, blobberID string) ([]AllocationBlobberTerm, error) {
+	var terms []AllocationBlobberTerm
+	return terms, edb.Store.Get().Model(&AllocationBlobberTerm{}).
+		Where(AllocationBlobberTerm{AllocationID: allocationID, BlobberID: blobberID}).
+		Find(&terms).Error
+}
+
+func (edb *EventDb) deleteAllocationBlobberTerm(allocationID string, blobberID string) error {
+	return edb.Store.Get().Model(&AllocationBlobberTerm{}).
+		Where("allocation_id = ? and blobber_id = ?", allocationID, blobberID).
+		Delete(&AllocationBlobberTerm{}).Error
+}
+
+func (edb *EventDb) deleteAllocationBlobberTerms(terms []AllocationBlobberTerm) error {
+	for _, t := range terms {
+		err := edb.deleteAllocationBlobberTerm(t.AllocationID, t.BlobberID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (edb *EventDb) updateAllocationBlobberTerm(term AllocationBlobberTerm) error {
+	exists, err := term.exists(edb)
```

Comment:
not sure that we should check it here, we can execute update and then check number of rows updated

Reply:
I think the idea is reading is much less taxing on the database that write operations. It's how update and addOrOverwrite methods have been handled in eventsDB in general.

Reply:
I meant slightly different thing, usually you execute update and then check if any rows were affected, gorm can do it like this
```
result.RowsAffected
```
https://gorm.io/docs/update.html#Updated-Records-Count

the benefit is that you execute single query and it works exactly the same as `if exist then update` logic

Reply:
result.RowsAffected is a post-op data, it's available only when the query has been executed, therefore it'd be a write operation. So I think the idea to use existing data was added to reduce write ops. But I think considering our use case, updating and then checking the result.RowsAffected would be a better implementation for most tables.

But for this case, where we have batch data, I think we should go for batch upsert, a single operation to update all the data. I am sending in a commit, ptal.